ArchitectureFlow/ArchitectureFlow/Login/Presentation/LoginPresenter.m

```diff
@@ -0,0 +1,65 @@
+
+#import "LoginPresenter.h"
+#import "Login.h"
+#import "LoginView.h"
+#import "LoginViewModel.h"
+#import "LoginRequestModel.h"
+#import "LoginOutput.h"
+#import "Promise.h"
+
+@interface LoginPresenter () <LoginOutput>
```

Comment:
Is Output same as Delegate? Should we just call it LoginDelegate or LoginUsecaseDelegate?

Reply:
It's a Clean Arch convention, please rewatch my presentaton we I talk about this into a good detail i think. The output of an interactor is not a callback, listener or a delegate.
When a caller calls interactor it asks it to do some job. So call an interactor the caller does 2 things:

1) It receives a receipt of this particular call (a Promise) that can tell the caller if this particular exection is sill going and gives a choise to cancel it. The Promise is not a "response" and is not where we get the "response" from. In fact there is no such a things as "response" from a use case, which is why there is no Callback/Listener/Delegate in traditional sense

2) It must comply with the Interactor's contract to be able to call it. The contract comes in 2 forms:
a) Request object: use case needs some data to be able to perform its execution (some might not need any of course). This is a POJO/POCO with a bunch of primitives or other data structures. (we should decide how many parameters should mean a Request class)
b) The Output. Again, I thought would be best to rewatch prsentation. If we look at the Clean Architecture diagram, below InpurBoundary we can see OutputBoundary:
![image](https://user-images.githubusercontent.com/23355816/36193959-c2f860e8-115f-11e8-8870-37d8273b8eeb.png)
This is the Protocol we are talking about here. This is not a simple success/failure listener, it's a contract. It tells the caller, down to necessary detail what operations it will need to run as a consequence of executing the interactor. For example, the output can have multiple methods on them and they can be called multiple times. Interactor can give data to the output, but it musn't return a result (Interactor cannot get extra data from Output). It's an Output so things only go outwards. It informs the caller about something, such as: validationSuccessful() while logging in. Or multiple different kind or errors. Or multiple different kind of data from some success flow (like loggedInSuccessfully(session). 
The reason why these are important as that interactor is not simply a layer that abstracts some operations. It's a layer that contains the entirety of the business logic that app has. That includes exclaiming various cases that it might result in. It can have multiple "successes" and "errors" methods if we want to call them that, but essentially they are part of the contract.